sycl/include/sycl/types.hpp

```diff
@@ -552,13 +523,15 @@ template <typename Type, int NumElements> class vec {
       std::is_fundamental_v<vec_data_t<Ty>> ||
           detail::is_half_or_bf16_v<typename std::remove_const_t<Ty>>,
       vec &>
+  // TODO: Can we use std::fill or something similar to vectorize assignment
+  // operation?
```

Comment:
I don't think STL handles vectorization and I'd expect auto-vectorizer to work in either scenario. Also, it's unclear how this differs between host/device code. I think you could re-arrange the code to make it clearer.

What might work is something like

```
#if VECTOR_EXT_AVAILABLE
return bit_cast<T>(bit_cast<ext_vector_t>(Op0) `OP` bit_cast<ext_vector_t>(Op1));
#else
for(...) { ... }
#endif
```

That would work for both host and device (if OP is natively supported in LLVM IR) and could simplify some code below. I'm not 100% sure though.